docs/reddit-migration/strategy/schema.md

```diff
@@ -0,0 +1,99 @@
+    - We still need to keep the `text` field and the `article` relation
+
+
+### Comments
+
+Comments should have the fields carried over with the following adjustments:
+
+ - `channel_id` -> `id`
+ - `reports_ignored = BooleanField(default=False)`
+   - Flag to indicate whether we're ignoring subsequent reports of the comment
+   - See "Moderation" for more details
+
+#### Comment Tree
+
+Prior art:
+
+ - SQL Antipatterns ("Naive Trees"):http://www.r-5.org/files/books/computers/languages/sql/style/Bill_Karwin-SQL_Antipatterns-EN.pdf
+ - Models for Hierarchical Data: https://www.slideshare.net/billkarwin/models-for-hierarchical-data
+ - Django treebeard: https://django-treebeard.readthedocs.io/en/latest/
+
+Reddit stores comment trees in Cassandra, which is overkill for our current use cases. That said, we need to be able to migrate the comments in such a manner that we achieve the following:
+
+- Maintain current API result structures
+  - Specifically, we need to support some form of reddit's "more comments" pagination. This **does not** need to partition the comments exactly as reddit would, the interface just needs to be compatible with how the frontend traverses it.
+- Support currently used sorting methods (best, old, new)
+- Support rendering an arbitrary subtree of the comments for the comment detail view
+
+From the Prior Art listed above, Materialized Paths are probably our best bet and we're already familiar to some degree with them because they're what `wagtail` uses under the hood for page hierarchies.
+
+The big thing here is that in order to support the sorts we need one for *each type of sort*. That precludes using a single table for everything so we'll want to have a hierarchical table for each sort and then `prefetch_related` on the `Comment` table so that code can work generically over similarly structured tables. This approach means these sorting hierarchy tables end up being a sorted lookup table for extracting a portion of the comment tree.
+
+### Votes
+
+We should have a table for `PostVote` and `CommentVote`, these should have the following columns:
+
+- `comment = ForeignKey(Comment)` / `post = ForeignKey(Post)` - item that was voted on
+- `voter = ForeignKey(User)` - the user who made the vote
+- `direction = SmallIntegerField()` - we store either a `1` or `-1` to indicate which direction the user voted. In practice, `PostVote` will always have a `direction == 1`, but it's useful to carve this out now in case that changes at a later date. It's also useful to be able to run a `SUM()` over that column.
+
+Votes are probably the trickiest moving piece in this reddit migration because it's impractical to get the data directly from the reddit backend for a backpopulate. This probably means we're going to need either some kind of ETL process. This is only a 1-time event, so an approach as basic as an export/import of a CSV would be adequate.
```

Comment:
sincere question: is it worth it to recover all of this data?

When I look at the production instance of open there is really not a lot of engagement going on right now. Many channels have a few posts at most, most with no comments and only one or two upvotes. some are more active, but from the channels I have access to it looks like the only ones that have semi-active engagement on them are the ones affiliated with micromasters programs (dedp and so on). even there, we have a post every couple of days. It doesn't seem like we have a lot of active users who are regularly posting, upvoting, etc, other than people who work in OL.

What if instead of spending a lot of time and effort getting every single vote from reddit we did the following:

- implement immediately a new vote table in postgres, and save all votes going forward in that table as well as in reddit
- at the time of migration, when we're ready to start calculating our own scores, do a backfill where we look at all the votes we have stored vs the scores for posts, comments, etc and create dummy votes for any differences. so like if there's a post which has a score of 10 but we have 6 upvotes in postgres then just create 4 upvotes with no user linked to them. that will keep the score the same for the post when we switch over to our new system.

My assumption, which could be wrong, is that switching off reddit completely is, even if we start working immediately, something which is going to take a pretty long time. So I was thinking that if we took the approach above, and starting saving votes ourselves immediately, by the time we're actually in a position to switch reddit off we'll probably already have accurate saved votes for the past four-six months of content. This means that as we switch to our own way to calculate score and show whether the user has upvoted content or not we'll have that data to use for recent content which the user is likely to see.

my thought is that we should at least ask whether showing a personalized upvote button to a user on old content is something that is worth spending a lot of engineering time on.

Anyhow - I put this out as food for thought, I'm not convinced one way or the other but I felt it would be worthwhile to raise the question.

I do think that probably in any case we may want to look into getting our own table going for recording votes soon, since it may make our job easier going forward.

Reply:
That's a fair question to ask, I don't think they're mutually exclusive either (we could take your suggestion initially and then remove and backpopulate later on as long as we have a backup of the original data). I defer to @pdpinch on what makes sense product/user-wise